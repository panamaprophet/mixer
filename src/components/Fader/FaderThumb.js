import React from 'react';


const FaderThumb = ({
	position = 0, 
	isVertical = false,
	onMoveStart = () => {},
	onMove = () => {},
	onMoveEnd = () => {},
}) => {
	const styleProperty = isVertical ? 'bottom' : 'left';
	const stylePropertyValue = position + 'px';

	return (
		<div
			className="fader-thumb"
			style={{[styleProperty]: stylePropertyValue}}
			onMouseMove={onMove}
			onMouseUp={onMoveEnd}
			onMouseDown={onMoveStart}
		>
		</div>
	);
};

export default FaderThumb;