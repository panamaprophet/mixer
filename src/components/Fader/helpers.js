'use strict';


import {min, max} from 'ramda';


/**
 * @typedef {Object} PointerPosition
 * @property {number} value - absolute value in pixels
 * @property {number} percentage - relative value in percents
 */


/**
 * @returns {number} vertical coordinate from touch or mouse event
 */
export const getY = event => event.touches ? event.touches[0].pageY : event.pageY;

/**
 * @returns {number} horizontal coordinate from touch or move event
 */
export const getX = event => event.touches ? event.touches[0].pageX : event.pageX;

/**
 * @returns {PointerPosition}
 */
export const getPointerVerticalPosition = (position, {top, bottom, height}) => {
    const value = bottom - min(min(position, top), bottom);

    return {
        value,
        percentage: value / (height / 100),
    };
};

/**
 * @returns {PointerPosition}
 */
export const getPointerHorizontalPosition = (position, {width, left}) => {
    const value = max(min(position, left + width), left) - left;

    return {
        value,
        percentage: value / (width / 100),
    }
};