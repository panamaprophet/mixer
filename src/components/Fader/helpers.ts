'use strict';


import {min, max} from 'ramda';

/**
 * @returns {number} vertical coordinate from touch or mouse event
 */
export const getY = event => event.touches ? event.touches[0].pageY : event.pageY;

/**
 * @returns {number} horizontal coordinate from touch or move event
 */
export const getX = event => event.touches ? event.touches[0].pageX : event.pageX;

/**
 * @returns {number} relative position in percents
 */
export const getPointerVerticalPosition = (position, {top, bottom, height}) => {
    let value = position;

    if (value < top) value = top;
    if (value > bottom) value = bottom;

    value = bottom - value;

    return value / (height / 100);
};

/**
 * @returns {number} relative position in percents
 */
export const getPointerHorizontalPosition = (position, {width, left}) => {
    let value = position; // max(min(position, left + width), left) - left;

    const rightBorder = left + width;

    if (value > rightBorder) value = rightBorder;
    if (value < left) value = left;

    value = value - left;

    return value / (width / 100);
};