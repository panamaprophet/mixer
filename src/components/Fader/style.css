.fader {
    height: 100%;

    text-align: center;

    background: #c4cace; /* url("i/metal-texture-small.jpg"); */
    background-size: contain;
}

.fader:not(:last-child) {
    margin-right: 1px;
}

.control {
    position: relative;

    height: calc(100% - 68px);
    margin-top: 34px;
    margin-bottom: 34px;
}

.control:after {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 50%;

    width: 10px;
    margin-left: -5px;

    content: '';

    background: #555;
    box-shadow: inset 0 0 10px #000;
}

.thumb {
    position: absolute;
    z-index: 10;
    bottom: 0;
    left: 50%;

    width: 30px;
    height: 60px;
    margin-left: -15px;

    cursor: s-resize;
    transform: translateY(50%);

    border-radius: 2px;
    background: linear-gradient(to bottom, #999 0%, #eee 11%, #ccc 48.33%, #333 50%, #ccc 51.66%, #eee 88.33%, #999 100%);
    box-shadow: 0 4px 5px rgba(0,0,0,.3), 0 2px 1px rgba(0,0,0,.5);

    touch-action: none;
}



.isHorisontal .control {
    margin: 0;
    padding: 0 15px;

    user-select: none;
}

.isHorisontal .thumb {
    top: -3px;
    left: auto;

    width: 30px;
    height: 15px;
    margin: 0;

    cursor: e-resize;
    transform: translateX(-50%);

    background: linear-gradient(to right, #999 0%, #eee 11%, #ccc 48.33%, #333 50%, #ccc 51.66%, #eee 88.33%, #999 100%);
}

.isHorisontal .control:after {
    left: 0;

    width: 100%;
    height: 10px;
    margin: 0;
}

.isAnimated {
    transition: all 250ms ease-in-out;
}
