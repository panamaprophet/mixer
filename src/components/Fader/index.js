'use strict';


import React, {useRef} from 'react';

import FaderThumb from './FaderThumb';

import {
    getX,
    getY,
    getPointerVerticalPosition,
    getPointerHorizontalPosition,
} from './helpers';


const Fader = ({
    position = 0,
    isVertical = false, 
    onChange = () => {},
}) => {
    const containerRef = useRef(null);

    const onMoveStart = event => {
        event.preventDefault();

        document.documentElement.addEventListener('mousemove', onMove);
        document.documentElement.addEventListener('mouseup', onMoveEnd);
    }

    const onMove = event => {
        event.preventDefault();

        const containerElement = containerRef.current;
        const offset = containerElement.getBoundingClientRect();

        const position = isVertical
            ? getPointerVerticalPosition(getY(event), offset)
            : getPointerHorizontalPosition(getX(event), offset);

        onChange(position);
    }

    const onMoveEnd = event => {
        event.preventDefault();

        document.documentElement.removeEventListener('mousemove', onMove);
        document.documentElement.removeEventListener('mouseup', onMoveEnd);
    }

    const className = ['fader', isVertical ? 'fader-vertical' : 'fader-horizontal'].join(' ');

    return (
        <div className={className} ref={containerRef}>
            <FaderThumb position={position} events={{onMouseDown: onMoveStart}} isVertical={isVertical} />
        </div>
    );
};

export default Fader;