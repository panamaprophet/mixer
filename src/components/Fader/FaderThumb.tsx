import React from 'react';
import style from './style.css';


type Props = {
    position: number,
    isVertical: boolean,
    events: Record<string, (event: TouchEvent | MouseEvent) => void>,
};


export const FaderThumb: React.FC<Props> = ({
    position = 0,
    isVertical = false,
    events = {},
}) => {
    const styleProperty = isVertical ? 'bottom' : 'left';
    const stylePropertyValue = `${position}%`;
    const inlineStyle = {[styleProperty]: stylePropertyValue};

    return (
        <div
            className={style.thumb}
            style={inlineStyle}
            {...events}
        >
        </div>
    );
}