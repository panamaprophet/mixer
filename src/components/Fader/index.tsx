import React, {useRef} from 'react';
import classnames from 'classnames';
import {FaderThumb} from './FaderThumb';
import {getX, getY, getPointerVerticalPosition, getPointerHorizontalPosition} from './helpers';
import {SendParameter, SendParameterType} from '/helpers/entities';
import style from './style.css';


type FaderEvent = 'mousemove' | 'mouseup' | 'mousedown' | 'touchmove' | 'touchend' | 'touchstart';

const EVENTS_MAP: Record<string, FaderEvent> = {
    'mousemove': 'touchmove',
    'mouseup': 'touchend',
    'mousedown': 'touchstart',
};

const hasTouchEventsSupport = (): boolean => 'ontouchstart' in window;

const getEventNameByFeature = (eventName: FaderEvent): FaderEvent => (hasTouchEventsSupport() ? EVENTS_MAP[eventName] : eventName);

interface Props extends Partial<SendParameter> {
    onChange: (value: number) => void,
    isVertical?: boolean,
}

export const Fader: React.FC<Props> = ({
    type = SendParameterType.FADER,
    onChange,
    isVertical = false,
    value,
}) => {
    if (type !== SendParameterType.FADER) {
        return null;
    }

    const containerRef = useRef<HTMLDivElement>(null);

    const onMoveStart = (event: MouseEvent | TouchEvent): void => {
        event.preventDefault();

        document.documentElement.addEventListener(getEventNameByFeature('mousemove'), onMove);
        document.documentElement.addEventListener(getEventNameByFeature('mouseup'), onMoveEnd);
    }

    const onMove = (event: MouseEvent | TouchEvent): void => {
        event.preventDefault();

        const containerElement = containerRef.current;

        if (containerElement) {
            const offset = containerElement.getBoundingClientRect();
            const x = getX(event) - document.documentElement.scrollLeft;
            const y = getY(event) - document.documentElement.scrollTop;

            const newValue = isVertical
                ? getPointerVerticalPosition(y, offset)
                : getPointerHorizontalPosition(x, offset);

            onChange(newValue);
        }
    }

    const onMoveEnd = (event: MouseEvent | TouchEvent): void => {
        event.preventDefault();

        document.documentElement.removeEventListener(getEventNameByFeature('mousemove'), onMove);
        document.documentElement.removeEventListener(getEventNameByFeature('mouseup'), onMoveEnd);
    }

    const thumbEventName = hasTouchEventsSupport() ? 'onTouchStart' : 'onMouseDown';

    return (
        <div className={classnames(style.fader, !isVertical && style.isHorisontal)} ref={containerRef}>
            <div className={style.control}>
                <FaderThumb position={Number(value)} events={{[thumbEventName]: onMoveStart}} isVertical={isVertical} />
            </div>
        </div>
    );
};