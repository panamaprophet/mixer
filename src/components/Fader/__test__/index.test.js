import React from 'react';
import {shallow, mount} from 'enzyme';

import Fader from '../';


describe('<Fader />', () => {
    it('should renders without any errors', () => {
        const wrapper = shallow(
            <Fader />
        );

        expect(wrapper).toMatchSnapshot();
    });

    describe('position', () => {
        it('should be passed to thumb', () => {
            const position = 42.0;
            const wrapper = shallow(
                <Fader position={position} />  
            );

            const result = wrapper.find('FaderThumb').prop('position');

            expect(result).toBe(position);
        });
    });

    describe('isVertical', () => {
        it('should have isHorisontal class on false', () => {
            const wrapper = shallow(
                <Fader isVertical={false} />
            );

            const result = wrapper.hasClass('isHorisontal');

            expect(result).toBe(true);
        });

        it('should omit isHorisontal class on true', () => {
            const wrapper = shallow(
                <Fader isVertical={true} />
            );

            const result = wrapper.hasClass('isHorisontal');

            expect(result).toBe(false);
        });
    });

    describe('onChange', () => {
        it('should be called on position change', () => {
            const onChangeMock = jest.fn();
            const wrapper = mount(
                <Fader onChange={onChangeMock} />
            );

            wrapper.find('FaderThumb').simulate('mouseDown');

            window.document.documentElement.dispatchEvent(new Event('mousemove'));
            window.document.documentElement.dispatchEvent(new Event('mouseup'));

            expect(onChangeMock).toBeCalled();
        });
    });
});