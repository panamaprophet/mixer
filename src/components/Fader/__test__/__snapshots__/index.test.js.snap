// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fader /> should renders without any errors 1`] = `
<div
  className="fader isHorisontal"
>
  <div
    className="control"
  >
    <FaderThumb
      events={
        Object {
          "onMouseDown": [Function],
        }
      }
      isVertical={false}
      position={0}
    />
  </div>
</div>
`;
